### PR TITLE
bpo-38565: add new cache_parameters method for lru_cache

### PR DESCRIPTION
**Pull request by Zheaoli**
https://bugs.python.org/issue38565

https://bugs.python.org/issue38565

### CODE REVIEW

File: Lib/functools.py [hunk 1]
```diff
@@ -613,6 +613,9 @@ def cache_info():
         with lock:
             return _CacheInfo(hits, misses, maxsize, cache_len())
 
+    def cache_parameters():
+        return {"maxsize": maxsize, "typed": typed}
```
Comment (rhettinger) [inline Lib/functools.py]: NIt:  Please use single quotes, the same as used by ``dict.__repr__()``

Comment (rhettinger) [reply Lib/functools.py]: Also move the function to just after *cache_clear()*.

File: Lib/test/test_functools.py [hunk 2]
```diff
@@ -1654,6 +1654,12 @@ def orig(x, y):
             with self.subTest(func=f):
                 f_copy = copy.deepcopy(f)
                 self.assertIs(f_copy, f)
+    def test_lru_cache_parameters(self):
```
Comment (rhettinger) [inline Lib/test/test_functools.py:1658]: Add a blank line before the *def*.

File: Lib/test/test_functools.py [hunk 3]
```diff
@@ -1654,6 +1654,12 @@ def orig(x, y):
...
+        def f():
+            return 1
+        f = self.module.lru_cache(2)(f)
```
Comment (rhettinger) [inline Lib/test/test_functools.py]: Let's use the decorator notation to create the example.  Also, make the example use `maxsize=1000` and `typed=True`.

File: Misc/NEWS.d/next/Library/2019-10-24-08-10-30.bpo-38565.SWSUst.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+add new cache_parameters method for lru_cache
```
Comment (rhettinger) [inline Misc/NEWS.d/next/Library/2019-10-24-08-10-30.bpo-38565.SWSUst.rst]: ``Add new cache_parameters() method for functools.lru_cache() to better support pickling.

File: Modules/_functoolsmodule.c [hunk 5]
```diff
@@ -1257,6 +1257,15 @@ lru_cache_cache_info(lru_cache_object *self, PyObject *unused)
                                  PyDict_GET_SIZE(self->cache));
 }
 
+static PyObject *
+lru_cache_cache_parameters(lru_cache_object *self)
+{
+    PyObject *cache_parameters =  PyDict_New();
```
Comment (rhettinger) [inline Modules/_functoolsmodule.c]: Need to test whether *cache_parameters* is *NULL*.

File: Modules/_functoolsmodule.c [hunk 6]
```diff
@@ -1257,6 +1257,15 @@ lru_cache_cache_info(lru_cache_object *self, PyObject *unused)
...
+    PyDict_SetItemString(cache_parameters, "maxsize", PyLong_FromSsize_t(self->maxsize));
```
Comment (rhettinger) [inline Modules/_functoolsmodule.c]: There are also two possible failure points that need to be tested here as well.

File: Modules/_functoolsmodule.c [hunk 7]
```diff
@@ -1257,6 +1257,15 @@ lru_cache_cache_info(lru_cache_object *self, PyObject *unused)
...
+    PyDict_SetItemString(cache_parameters, "typed", self->typed == 0 ? Py_False : Py_True);
```
Comment (rhettinger) [inline Modules/_functoolsmodule.c]: 1.  Drop the test for zero, putting the positive case first: ``self->typed ? Py_True : Py_False)``
2. Test the result of *PyDict_SetItemString()* to see if it failed.

File: Modules/_functoolsmodule.c [hunk 8]
```diff
@@ -1335,6 +1344,7 @@ cache_info_type:    namedtuple class with the fields:\n\
 
 static PyMethodDef lru_cache_methods[] = {
     {"cache_info", (PyCFunction)lru_cache_cache_info, METH_NOARGS},
+    {"cache_parameters", (PyCFunction)lru_cache_cache_parameters, METH_NOARGS},
```
Comment (rhettinger) [inline Modules/_functoolsmodule.c]: Move this down one, just after the *cache_clear* entry.

**Review by rhettinger — CHANGES_REQUESTED** [review]

**Comment by Zheaoli** [comment]
ping @rhettinger

**Review by rhettinger — APPROVED** [review]

**Pull request merged**